OEP-6/OEP-6.mediawiki

```diff
@@ -0,0 +1,405 @@
+
+Returns Description object of the <code>identity</code>. This object contains public keys and attributes of the identity.
+
+* Rejects with '''MALFORMED_IDENTITY''' in case the <code>identity</code> is not a valid identity
+
+====addAttributes====
+<pre>
+function addAttributes({ attributes: OntIdAttribute[] }): Promise<void>
+</pre>
+
+Initiates adding of attributes to the user identity. This action needs to be confirmed by the user. Attributes with existing <code>key</code> will be overwritten.
+
+* Rejects with '''NO_IDENTITY''' in case the user is not signed in or has no identity
+
+====removeAttribute====
+<pre>
+function removeAttribute({ key: string }): Promise<void>
+</pre>
+
+Initiates removing of attribute from the user identity. This action needs to be confirmed by the user.
+
+* Rejects with '''NO_IDENTITY''' in case the user is not signed in or has no identity
+
+
+===Message===
+This API deals with arbitrary message signing and verification.
+
+====signMessageHash====
+<pre>
+
+interface Signature {
+  publicKey: string;
+  data: string;
+}
+
+function signMessageHash({ messageHash: string }): Promise<Signature>
+</pre>
+
+Initiates signing of arbitrary <code>messageHash</code> by the user account or identity. dAPI provider should provide the user with a way for selecting which account or identity will sign the message.
+
+Returns the <code>Signature</code> object containing both the <code>publicKey</code> and signature <code>data</code> needed for signature validation.
+
+This method does not require the dApp to disclose the message itself, only the hash. 
+
+Because malicious dApp can hash a real prepared transfer transaction and plant it for signing, that posses a risk to the user.
+Therefore the hash is prepended with known string '''Ontology message:''', and only this hash is put for signing.
+
+* Rejects with '''NO_ADDRESS''' in case the user is not signed in or has no account or identity
+* Rejects with '''MALFORMED_MESSAGE''' in case the <code>message</code> is not hex encoded
+
+====verifyMessageHash====
+<pre>
+function verifyMessageHash({ messageHash: string, signature: Signature }): Promise<boolean>
+</pre>
+
+Verifies that the <code>signature</code> is valid signature of <code>messageHash</code>. 
+
+This method does not require the dApp to disclose the message itself, only the hash. The hash is prepended with known string '''Ontology message:''' and only this hash is put for verification.
+
+* Rejects with '''MALFORMED_MESSAGE''' in case the <code>message</code> is not hex encoded
+* Rejects with '''MALFORMED_SIGNATURE''' in case the <code>signature</code> is not in valid format
+
+====signMessage====
+<pre>
+function signMessage({ message: string }): Promise<Signature>
+</pre>
+
+Initiates signing of arbitrary <code>message</code> by the user account or identity. dAPI provider should provide the user with a way for selecting which account or identity will sign the message.
```

Comment:
```suggestion
Initiates signing of arbitrary <code>message</code> by the user account or identity. dAPI provider should provide the user with a way for selecting which account or identity will sign the message.
```
I think we should add another parameter in the function to decide the signer is user account of identity. If we let the dApi provider to the selection, the caller won't know what the signer is .

So the function should be like :
```
function signMessage({ message: string, signer: string})
```
`signer` can be value of `account` or `identity`. If this parameter is not provided, the function will use account to sign as default.